src/decompress.c

```diff
@@ -108,22 +254,797 @@ void LoadCompressedSpritePaletteOverrideBuffer(const struct CompressedSpritePale
+                loVec += 4;
+            }
+            else
+            {
+                loVec += 3;
+            }
+        }
+        else
+        {
+            currLength = loVec[0] & FIRST_LO_MASK;
+            currOffset = loVec[1] & FIRST_LO_MASK;
+
+            if (loVec[1] & CONTINUE_BIT)
+            {
+                currOffset |= (loVec[2] << 7);
+                loVec += 3;
+            }
+            else
+            {
+                loVec += 2;
+            }
+        }
+
+        if (currLength != 0)
+        {
+            u16 symVecVal = *symVec;
+            *dest++ = *symVec++;
+            if (currOffset == 1)
+            {
+                Fill16(symVecVal, dest, currLength);
+                dest += currLength;
+            }
+            else
+            {
+                // Copy16 is slower in this case.
+                u16 *from = dest - currOffset;
+                u16 *to = dest + currLength;
+                do {
+                    *dest++ = *from++;
+                } while (dest != to);
+            }
+        }
+        else
+        {
+            Copy16(symVec, dest, currOffset);
+            dest += currOffset;
+            symVec += currOffset;
+        }
+    } while (loVec < loVecEnd);
+}
+
+//  Dark Egg magic
+ARM_FUNC __attribute__((no_reorder)) static void SwitchToArmCallDecodeInstructions(u32 headerLoSize, const u8 *loVec, const u16 *symVec, void *dest, void (*decodeFunction)(u32 headerLoSize, const u8 *loVec, const u16 *symVec, void *dest))
+{
+    decodeFunction(headerLoSize, loVec, symVec, dest);
+}
+
+//  Dark Egg magic
+static void DecodeInstructionsIwram(u32 headerLoSize, const u8 *loVec, const u16 *symVec, void *dest)
+{
+    u32 funcBuffer[350];
+
+    CopyFuncToIwram(funcBuffer, DecodeInstructions, SwitchToArmCallDecodeInstructions);
+    SwitchToArmCallDecodeInstructions(headerLoSize, loVec, symVec, dest, (void *) funcBuffer);
+}
+
+//  Entrance point for smol compressed data
+void SmolDecompressData(const struct SmolHeader *header, const u32 *data, void *dest)
+{
+    //  This is apparently needed due to Game Freak sending bullshit down the decompression pipeline
+    if (header->loSize == 0 || header->symSize == 0)
+        return;
+    const u8 *leftoverPos = (u8 *)data;
+
+    sCurrState = header->initialState;
+
+    u32 headerLoSize = header->loSize;
+    u32 headerSymSize = header->symSize;
+    u32 alignedLoSize = header->loSize % 2 == 1 ? headerLoSize + 1 : headerLoSize;
+    u32 alignedSymSize = header->symSize % 2 == 1 ? headerSymSize + 1 : headerSymSize;
+    // Everything needs to be aligned.
+    void *memoryAlloced = Alloc((alignedSymSize*2) + alignedLoSize);
+    u16 *symVec = memoryAlloced;
+    u8 *loVec = memoryAlloced + alignedSymSize*2;
+    bool32 loEncoded = isModeLoEncoded(header->mode);
+    bool32 symEncoded = isModeSymEncoded(header->mode);
+    bool32 symDelta = isModeSymDelta(header->mode);
+
+    const u32 *pLoFreqs;
```

Comment:
Some compilers may warn about a potentially unused variable, so I would just assign `= NULL;` to prevent compiler errors.